rfcs/2021-06-01-multilinks.md

```diff
@@ -0,0 +1,354 @@
+```
+
+You can also have multiple URLs and a label:
+
+```
+& http://example.com|ipfs://example|dat://example Example label
+```
+
+# Reference-level explanation
+
+Link blocks start with `& `. They reference other files within the flow of a Subtext document. Any kind of file can be linked, including other Subtext documents.
+
+Links are the most important feature in Subtext. By allowing you to reference other documents, you can compose hypertext documents from many smaller documents.
+
+Links can include zero or more pipe-delimited URLs, and an optional text lable. They often take one of the following forms:
+
+- URL
+- URL with label
+- Multiple URLs
+- Multiple URLs with label
+- Label only (wikilink)
+
+The following are all valid URLs
+
+```
+& example.png
+& http://example.com
+& http://example.com Example label
+& http://example.com|ipfs://example|dat://example Example label
+& Example label
+```
+
+The client MUST render some kind of user-interactable link to the document. The client MAY choose the form this link takes.
+
+When more than one URL is present, clients MAY choose which URL to use when presenting to the user, when navigating, or when transcluding content.
+
+When no URL is present, but a text lable is present, clients SHOULD choose a resource that best matches the text when navigating to the link, or when transcluding content.
+
+For file types it supports, the client MAY render all or part of the linked document in-place (e.g. transclude). For example:
+
+- A linked image file (`.png`, `.jpg`, `.gif`, etc) MAY be rendered in-place by the client and sized to fit its context.
+- A linked video file MAY be rendered in-place, together with playback controls.
+- A linked `.csv` file MAY be rendered in-place as a table.
+- A linked Subtext file MAY be transcluded (linked while rendered in-place), rendered in-place in full, or excerpted and rendered in-place, depending on the use-case.
+
+Rather than extending the syntax of Subtext to include features like tables, videos, or deeply nested lists, our sense is that a hypertext format allows these special types to be represented in their native file containers. Clients that understand these other file types MAY embed them, or even allow you to edit them in-place. This keeps Subtext simple, and allows data sources like `.csv`, or `.png` to be OPTIONALLY embedded in-place, while remaining valid file types that can be opened, edited, and used in best-of-breed applications.
+
+## Syntax
+
+Each link form is some combination of zero or more pipe-delimited URLs, followed by optional label text.
+
+A link block has the following syntax in ABNF notation, as described in [IETF RFC 5234](https://tools.ietf.org/rfc/rfc5234):
+
+```abnf
+link-block = "&" *space link-block-body universal-newline
+link-block-body = url / labeled-url / url-group / labeled-url-group / text
+labeled-url = url 1*space text
+url-group = *(url "|") url
+labeled-url-group = url-group 1*space text
```

Comment:
Am I understanding correctly that white-space in URLs will have to be appropriately encoded by the author?

Reply:
Yes, but this is a requirement of URLs as specified in https://datatracker.ietf.org/doc/html/rfc3986

Reply:
Please note that a path can neither contain a space because everything after the first space is label. This is the right choice but might be worth a comment because files in file systems can have spaces in them.

Reply:
After consideration, I've modified the grammar so that it supports spaces in file paths. Here's how:

A link block may contain either a local file path with optional label, OR one or more URLs with optional label, OR just a label.

Local file paths MUST:

- Start with ./, ../, or /
- End with an extension

This allows us to disambiguate paths with spaces from labels. I think this is actually an improvement on the grammar, since you don't really need or want multiple local paths.

Also updated parser to implement this behavior.